File: src/test/java/org/roborace/lapscounter/service/StopwatchTest.java
package org.roborace.lapscounter.service;

import org.junit.jupiter.api.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.lessThanOrEqualTo;

class StopwatchTest {

    private static final long DELAY = 125L;

    @Test
    void testHappyPath() throws InterruptedException {
        Stopwatch stopwatch = new Stopwatch();
        assertThat(stopwatch.getTime(), is(0L));

        stopwatch.start();
        Thread.sleep(DELAY);
        assertThat(stopwatch.getTime(), lessThanOrEqualTo(DELAY + 10));

        stopwatch.finish();
        assertThat(stopwatch.getTime(), lessThanOrEqualTo(DELAY + 20));
    }

    @Test
    void testStartAfterFinish() throws InterruptedException {
        Stopwatch stopwatch = new Stopwatch();
        stopwatch.start();
        Thread.sleep(DELAY);
        assertThat(stopwatch.getTime(), lessThanOrEqualTo(DELAY + 10));

        stopwatch.finish();
        assertThat(stopwatch.getTime(), lessThanOrEqualTo(DELAY + 10));

        stopwatch.start();
        assertThat(stopwatch.getTime(), is(0L));
        Thread.sleep(DELAY);
        assertThat(stopwatch.getTime(), lessThanOrEqualTo(DELAY + 10));
    }

    @Test
    void testReset() throws InterruptedException {
        Stopwatch stopwatch = new Stopwatch();
        stopwatch.start();
        Thread.sleep(DELAY);
        assertThat(stopwatch.getTime(), lessThanOrEqualTo(DELAY + 10));

        stopwatch.finish();
        assertThat(stopwatch.getTime(), lessThanOrEqualTo(DELAY + 10));

        stopwatch.reset();
        Thread.sleep(DELAY);
        assertThat(stopwatch.getTime(), is(0L));
    }

}

File: src/test/kotlin/org/roborace/lapscounter/service/util/StopwatchTest.kt
package org.roborace.lapscounter.service.util

import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.greaterThanOrEqualTo
import org.hamcrest.Matchers.lessThanOrEqualTo
import org.junit.jupiter.api.Test

internal class StopwatchTest {
    @Test
    fun testHappyPath() {
        val stopwatch = Stopwatch()
            .assertTimeEqual(0L)

        stopwatch.start()
            .sleep(DELAY)
            .assertTimeEqualSmooth(DELAY)

        stopwatch.stop()
            .assertTimeEqualSmooth(DELAY)
    }

    @Test
    fun testTimeStillSameAfterFinish() {
        val stopwatch = Stopwatch()
            .start()
            .sleep(DELAY)
            .stop()
            .assertTimeEqualSmooth(DELAY)

        stopwatch.sleep(DELAY)
            .assertTimeEqualSmooth(DELAY)
    }

    @Test
    fun testStartAfterFinish() {
        val stopwatch = Stopwatch().start()
            .sleep(DELAY)
            .stop()
            .assertTimeEqualSmooth(DELAY)

        stopwatch.start()
            .assertTimeEqual(0L)
            .sleep(DELAY)
            .assertTimeEqualSmooth(DELAY)
    }

    @Test
    fun testReset() {
        val stopwatch = Stopwatch().start()
            .sleep(DELAY)
            .assertTimeEqualSmooth(DELAY)
            .stop()
            .assertTimeEqualSmooth(DELAY)

        stopwatch.reset()
            .sleep(DELAY)
            .assertTimeEqual(0L)
    }

    @Test
    fun testPauseContinue() {
        val stopwatch = Stopwatch().start()
            .sleep(DELAY)
            .stop()
            .sleep(DELAY)

        stopwatch.`continue`()
            .assertTimeEqualSmooth(DELAY)
            .sleep(DELAY)
            .assertTimeEqualSmooth(2 * DELAY)
    }

    @Test
    fun testPauseContinueFinish() {
        val stopwatch = Stopwatch().start()
            .sleep(DELAY)
            .stop()
            .sleep(DELAY)
            .`continue`()
            .sleep(DELAY)

        stopwatch.stop()
            .assertTimeEqualSmooth(2 * DELAY)
            .sleep(DELAY)
            .assertTimeEqualSmooth(2 * DELAY)
    }

    companion object {
        private const val DELAY = 125L
    }
}

private fun Stopwatch.sleep(delay: Long): Stopwatch {
    Thread.sleep(delay)
    return this
}

private fun Stopwatch.assertTimeEqual(ms: Long) = this.apply {
    assertThat(this.time(), `is`(ms))
}

private fun Stopwatch.assertTimeEqualSmooth(ms: Long) = this.apply {
    val time = this.time()
    assertThat(time, lessThanOrEqualTo(ms + 15))
    assertThat(time, greaterThanOrEqualTo(ms))
}
